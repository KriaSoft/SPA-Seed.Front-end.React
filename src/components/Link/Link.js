/**
 * React Starter Kit (https://www.reactstarterkit.com/)
 *
 * Copyright Â© 2014-2016 Kriasoft, LLC. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import React, { PropTypes } from 'react';
import history from '../../core/history';

function isLeftClickEvent(event) {
  return event.button === 0;
}

function isModifiedEvent(event) {
  return !!(event.metaKey || event.altKey || event.ctrlKey || event.shiftKey);
}

class Link extends React.Component {
  static propTypes = {
    to: PropTypes.string.isRequired,
    query: PropTypes.oneOfType([
      React.PropTypes.string,
      React.PropTypes.object,
    ]),
    state: PropTypes.object,
    children: PropTypes.node,
    onClick: PropTypes.func,
  };

  handleClick = (event) => {
    if (this.props.onClick) {
      this.props.onClick(event);
    }

    if (isModifiedEvent(event) || !isLeftClickEvent(event)) {
      return;
    }

    if (event.defaultPrevented === true) {
      return;
    }

    event.preventDefault();
    history.push({
      pathname: this.props.to,
      state: this.props.state ? this.props.state : undefined,
      search: this.props.query ? this.params(this.props.query) : undefined,
    });
  };

  params = (arg) => {
    let params;

    if (typeof arg === 'string') {
      params = arg.startsWith('?') ? arg : `?${arg}`;
    } else if (arg === Object(arg) && Object.prototype.toString.call(arg) === '[object Object]') {
      params = `?${Object.keys(this.props.query).map((key) => (
        `${encodeURIComponent(key)}=${encodeURIComponent(this.props.query[key])}`
      )).join('&')}`;
    }

    return params;
  }

  render() {
    // eslint-disable-next-line no-unused-vars
    const { to, children, query, state, ...props } = this.props;
    return <a href={to} {...props} onClick={this.handleClick}>{children}</a>;
  }
}

export default Link;
