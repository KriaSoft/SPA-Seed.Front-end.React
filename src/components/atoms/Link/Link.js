/**
 * React Starter Kit (https://www.reactstarterkit.com/)
 *
 * Copyright Â© 2014-present Kriasoft, LLC. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';
import history from '../../../history';

function isLeftClickEvent(event) {
  return event.button === 0;
}

function isModifiedEvent(event) {
  return !!(event.metaKey || event.altKey || event.ctrlKey || event.shiftKey);
}

/**
 * Link component for interal linking with the router.
 * If baseUrl is set (in config.js) all links are prefixed with the baseUrl
 */
class Link extends React.Component {
  static propTypes = {
    /** The internal link of the page where you linking to. */
    to: PropTypes.string.isRequired,
    /** The inner childs. */
    children: PropTypes.node.isRequired,
    /** Function that will be fired on click. */
    onClick: PropTypes.func,
  };

  static contextTypes = {
    baseUrl: PropTypes.string.isRequired,
  };

  static defaultProps = {
    onClick: null,
  };

  handleClick = event => {
    const { baseUrl } = this.context;
    const { to, onClick } = this.props;

    if (onClick) {
      onClick(event);
    }

    if (isModifiedEvent(event) || !isLeftClickEvent(event)) {
      return;
    }

    if (event.defaultPrevented === true) {
      return;
    }

    event.preventDefault();
    history.push(`${baseUrl}${to}`);
  };

  render() {
    const { baseUrl } = this.context;
    const { children, to, ...props } = this.props;

    return (
      <a href={`${baseUrl}${to}`} {...props} onClick={this.handleClick}>
        {children}
      </a>
    );
  }
}

export default Link;
