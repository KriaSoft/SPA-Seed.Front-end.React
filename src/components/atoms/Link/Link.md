```js
<Link to="/">Home</Link>
```
