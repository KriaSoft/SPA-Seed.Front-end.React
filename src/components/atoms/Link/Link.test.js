/**
 * React Starter Kit (https://www.reactstarterkit.com/)
 *
 * Copyright Â© 2014-present Kriasoft, LLC. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

/* eslint-env jest */
/* eslint-disable padded-blocks, no-unused-expressions */

import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Link from './Link';

describe('<Link />', () => {
  const props = {
    to: '/internal-link',
    children: ['Internal Link'],
  };

  const wrapper = shallow(<Link {...props} />, {
    context: {
      baseUrl: '',
    },
  });

  test('renders href without baseUrl correctly', () => {
    expect(wrapper.find('a').prop('href')).toBe('/internal-link');
  });

  test('renders href with baseUrl correctly', () => {
    wrapper.setContext({ baseUrl: '/base-url' });
    expect(wrapper.find('a').prop('href')).toBe('/base-url/internal-link');
  });

  test('renders text correctly', () => {
    expect(wrapper.text()).toBe('Internal Link');
  });

  test('renders children correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
