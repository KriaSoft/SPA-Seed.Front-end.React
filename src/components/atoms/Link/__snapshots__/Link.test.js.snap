// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> renders children correctly 1`] = `
<a
  href="/base-url/internal-link"
  onClick={[Function]}
>
  Internal Link
</a>
`;
